OvmfPkg/PlatformPei/PlatformId.c

```diff
@@ -0,0 +1,660 @@
+ * span [Evts, Evts + EvtsSize). Associates the string data with
+ * gTcg800155PlatformIdEventHobGuid. Any unused bytes or out-of-bounds event
+ * sizes are considered corrupted and are discarded.
+**/
+STATIC
+VOID
+PlatformIdRegisterSp800155 (
+  IN CONST EFI_PEI_SERVICES  **PeiServices,
+  IN       UINT8             *Evts,
+  IN       UINTN             EvtsSize
+  )
+{
+  EFI_STATUS                               Status;
+  UINTN                                    EvtSize;
+  VOID                                     *Hob;
+  EFI_HOB_GUID_TYPE                        *GuidHob;
+  UINT8                                    *Evt;
+  UINT8                                    *EvtNext;
+  UINT8                                    *EvtsEnd;
+  UINT8                                    *EvtDest;
+  TCG_Sp800_155_PlatformId_Event3_Cursors  Data;
+
+  Evt     = Evts;
+  EvtsEnd = Evts + EvtsSize;
+  while (Evt != EvtsEnd) {
+    Status = ParseSp800155 (Evt, EvtsEnd, &Data);
+    if (EFI_ERROR (Status)) {
+      DEBUG ((
+        DEBUG_ERROR,
+        DPREFIX "malformed eventdata event in fw_cfg, skipping\n"
+        ));
+      return;
+    }
+
+    if (DebugPrintLevelEnabled (DEBUG_VERBOSE)) {
+      DebugCursors (DEBUG_VERBOSE, &Data);
+    }
+
+    EvtNext = Data.PlatformCertLocatorEnd;
+    if (EvtNext == NULL) {
+      // Must be an Event2 structure.
+      EvtNext = Data.FirmwareVersionEnd;
+    }
+
+    EvtSize = (UINTN)(EvtNext - Evt);
+    if (EvtSize > MAX_UINT16 - sizeof (EFI_HOB_GUID_TYPE)) {
+      DEBUG ((DEBUG_ERROR, DPREFIX "Eventdata too large for HOB, skipping\n"));
+      Evt = EvtNext;
+      continue;
+    }
+
+    Status = (*PeiServices)->CreateHob (
+                               PeiServices,
+                               EFI_HOB_TYPE_GUID_EXTENSION,
+                               sizeof (EFI_HOB_GUID_TYPE) + (UINT16)EvtSize,
+                               &Hob
+                               );
+    // If the proposed size is too much, ignore the host message and continue.
+    if (Status == EFI_SUCCESS) {
+      GuidHob = (EFI_HOB_GUID_TYPE *)Hob;
+      CopyGuid (&GuidHob->Name, &gTcg800155PlatformIdEventHobGuid);
+      EvtDest = (UINT8 *)GET_GUID_HOB_DATA (Hob);
+      CopyMem (EvtDest, Evt, EvtSize);
+      // Fill the remaining HOB padding bytes with 0s.
+      SetMem (EvtDest + EvtSize, GET_GUID_HOB_DATA_SIZE (Hob) - EvtSize, 0);
+    }
+
+    Evt = EvtNext;
+  }
+}
+
+VOID
+PlatformIdInitialization (
+  IN CONST EFI_PEI_SERVICES  **PeiServices
+  )
+{
+  EFI_STATUS            Status;
+  UINTN                 NumPages;
+  EFI_PHYSICAL_ADDRESS  Pages;
+  FIRMWARE_CONFIG_ITEM  FdtItem;
+  UINTN                 FdtSize;
+  UINT8                 *Evts;
+
+  //
+  // The format of opt/org.tianocode/sp800155evts is expected to be a sequence of
+  // TCG_Sp800_155_PlatformId_Event2 or TCG_Sp800_155_PlatformId_Event3 structs.
+  //
+  Status = QemuFwCfgFindFile ("opt/org.tianocode/sp800155evts", &FdtItem, &FdtSize);

```

Comment:
Do you think if it is possible to input sp800155 event one by one?

That can reduce the parsing effort in final production.
We can just take one event from VMM and translate to TCG Hob directly.

Reply:
Do you want it to be sp800155evts is a Count, and then we read files "opt/org.tianocode/sp800155evts/%d" with the index ranging [0, Count)?

Reply:
Yes. I like your above idea. That will simplify our firmware work.

Reply:
I've removed the count file since it's unnecessary. I just start from 0 and go until the index number path doesn't exist.

Reply:
I like this simpliciation.